### ISSUE TITLE
twisted.web2 tests don't work in cleantests-1883

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#1906]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)| @jml reported|
|-|-|
|Trac ID|trac#1906|
|Type|defect|
|Created|2006-07-07 11:58:38Z|

The twisted.web2 tests come up with _lots_ of clean up errors when run in cleantests-1883.  The tests don't clean up after themselves properly.

I had a stab at fixing them, but the errors are fairly deep (for example, t.web2.stream has a callLater that isn't always cleaned up).

I'll list the failing tests in a comment

<details><summary>Searchable metadata</summary>

```
trac-id__1906 1906
type__defect defect
reporter__jml jml
priority__highest highest
milestone__Web2_Gold_Master Web2-Gold-Master
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__web2 web2
keywords__tests tests
time__1152273518000000 1152273518000000
changetime__1171901073000000 1171901073000000
version__None None
owner__dreid dreid
cc__oubiwann cc__dried
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)|@jml commented|
|-|-|

```
twisted.web2.test.test_static.TestFileSaver.test_compareFileContents 
twisted.web2.test.test_static.TestFileSaver.test_reportFileSave
twisted.web2.test.test_stream.FallbackSplitTest.test_closeleft 
twisted.web2.test.test_xmlrpc.XMLRPCTestCase.testErrors
twisted.web2.test.test_xmlrpc.XMLRPCTestCase.testResults 
twisted.web2.test.test_xmlrpc.XMLRPCTestCase2.testErrors 
twisted.web2.test.test_xmlrpc.XMLRPCTestCase2.testResults 
twisted.web2.test.test_xmlrpc.XMLRPCTestIntrospection.testErrors 
twisted.web2.test.test_xmlrpc.XMLRPCTestIntrospection.testListMethods 
twisted.web2.test.test_xmlrpc.XMLRPCTestIntrospection.testMethodHelp 
twisted.web2.test.test_xmlrpc.XMLRPCTestIntrospection.testMethodSignature 
twisted.web2.test.test_xmlrpc.XMLRPCTestIntrospection.testResults 
```

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)|@jyknight commented|
|-|-|

This is w.r.t. [#1883](https://github.com/twisted/twisted/issues/1883), right? Does making the suggested change there fix these?

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)|@jml commented|
|-|-|

Yeah, this is wrt [[#1883](https://github.com/twisted/twisted/issues/1883)](https://github.com/twisted/twisted/issues/1883)

No matter what happens with to Trial's code in #1883, these tests still don't clean up after themselves.

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)|@dreid commented|
|-|-|

I got rid of FileSaver, which isn't an elegant solution but it's probably the best one.  And I found a lonely deferred in PostTruncatorStream which fixes twisted.web2.test.test_stream.FallbackSplitTest.test_closeleft.  I am however at a complete loss as to what to do about xmlrpc short of rewriting the tests to not use the network.  Which although is something I'd like to do ([#1929](https://github.com/twisted/twisted/issues/1929)) I think it would only prevent a legitimate bug from being noticed.

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)|@dreid commented|
|-|-|

branched to [branches/web2-cleantests-1906](branches/web2-cleantests-1906)

**Comment by twisted-trac** [comment]
|[<img alt="oubiwann's avatar" src="https://avatars.githubusercontent.com/u/894251?s=50" width="50" height="50">](https://github.com/oubiwann)<a name="note_7"></a>|@oubiwann set owner to @oubiwann<br>@oubiwann set status to `assigned`|
|-|-|

I was working on this last week and ran into some issues. Tonight I got a simplified test case working (single listen, multiple connects, multiple proper/complete disconnects). I'm planning on rewritting the ```web2.xmlrpc``` tests with this approach this week.

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)<a name="note_10"></a>|@dreid set owner to @dreid<br>@dreid set status to `new`|
|-|-|

The xmlrpc failures were fixed by [#1929](https://github.com/twisted/twisted/issues/1929).

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)<a name="note_11"></a>|@dreid removed owner|
|-|-|

I rewrote the CGI tests, and added docstrings for them.  My recomendation for FileSaver is still to just get rid of it.  I'll rewrite it later as a tutorial.  

branched to [branches/web2-cleantests-1906-2](branches/web2-cleantests-1906-2)

A build is pending on the win32er buildbot.

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)|@dreid set owner to @jml|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_14"></a>|@exarkun set owner to @dreid<br>@exarkun set status to `new`|
|-|-|

Trivial issues:

  * pyflakes reports a number of warnings for each of the 4 files changes in this branch
  * There's trailing whitespace in test_cgi
  * \ line continuations aren't actually explicitly forbidden by the coding standard (although I think they should be).  But the uses of this construct in test_cgi are superfluous anyway.  Please remove them.
  * What's the deal with test_cgi.PythonScript.filter?  It looks unused to me.
  * The sibpath usage at the end of test_cgi.CGI.setUpResource is pointless
  * test_cgi.CGI.testRealAllInput has a typo in its name, ditto for its callback
  * CGIDirectoryTest subclasses CGI now, which means it includes all of CGI's tests, which is pointless.  CGI's useful methods should be in a mixin which both CGI and CGIDirectoryTest include.

Thanks for adding docstrings to all those test methods.  Also, the refactoring to remove all the duplicate path mangling is great.

Less trivial:

  * PostTruncaterStream.read change needs a unit test
  * It's not clear to me that it is correct to simply remove FileSaver.  If I understand correctly, it is being removed because it is hard to test.  If it is hard for the authors to unit test code like this, how is anyone else going to manage it?  What options other than removing it do we have?

Since the test_cgi cleanups seem unrelated to the test_static cleanups, it might make sense to create another ticket for the test_static changes so that the test_cgi improvements aren't blocked by those.

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)<a name="note_15"></a>|@dreid commented|
|-|-|

I'm not sure the PostTruncatorStream.read change is actually correct.  And I have reservations about removing FileSaver now too.

How about I back out both of those changes and file tickets for them?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_16"></a>|@exarkun commented|
|-|-|

Yes?

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)<a name="note_17"></a>|@dreid commented|
|-|-|

I've filed [#2433](https://github.com/twisted/twisted/issues/2433) and [#2434](https://github.com/twisted/twisted/issues/2434) to deal with the PostTruncatorStream and the FileSaver errors.

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)<a name="note_18"></a>|@dreid removed owner|
|-|-|

FileSaver and PostTruncaterStream changes reverted.  Putting this back up for review.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_22"></a>|@exarkun set owner to @dreid|
|-|-|

Spelling error in the docstring for PythonScript, "it's" should be "its".  To be extra pedantic, that sentence is also missing a period. ;)

Formatting in test_cgi.py is pretty inconsistent, but it was already, so eh.  It would be nice if all the new docstrings looked like this, though:

```
"""
foo
"""
```

instead of like this:

```
"""foo
"""
```

Otherwise this looks fine to merge.

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)<a name="note_23"></a>|@dreid set status to `closed`|
|-|-|

(In [19665]) Merge web2-cleantests-1906-2

Author: dreid
Reviewer: exarkun
Fixes [#1906](https://github.com/twisted/twisted/issues/1906)

Improve twisted.web2.test.test_cgi by preventing it from using the network and 
dropping dependency on twisted.web.client, also rename test methods and improve
docstrings.
